libs/remote_storage/src/s3_bucket.rs

```diff
@@ -657,40 +657,67 @@ impl RemoteStorage for S3Bucket {
         let max_retries = 10;
         let is_permanent = |_e: &_| false;
 
-        let list = backoff::retry(
-            || async {
-                Ok(self
-                    .client
-                    .list_object_versions()
-                    .bucket(self.bucket_name.clone())
-                    .set_prefix(prefix.clone())
-                    .send()
-                    .await?)
-            },
-            is_permanent,
-            warn_threshold,
-            max_retries,
-            "listing object versions for time_travel_recover",
-            backoff::Cancel::new(cancel.clone(), || anyhow!("Cancelled")),
-        )
-        .await?;
+        let mut key_marker = None;
+        let mut version_id_marker = None;
+        let mut versions_deletes = Vec::new();
+
+        loop {
```

Comment:
could you give input onto how that would look like? i don't understand. The list of the versions/deletes has to be buffered one way or another because I'm not sure about the order in which they are returned: having a complete list of versions for each key is very important so that we can determine the latest version that the key had at the specific point in time.

Reply:
On a second thought this wouldn't work with the current implementation because you need the versions to be sorted by key. You could abstract this for loop into a `Stream` implementation with `Item=VerOrDelete` and modify the `vds_for_key` map to have an ordered set as its value. That's not worth it since you need to buffer anyway though.

Initially, I assumed that we'd get the versions ordered by key, but the documentation states otherwise:
```
returns objects in the order that they were stored, returning the most recently stored object first, starting with the value for key-marker
```

Feel free to resolve the comment.

Reply:
Yeah the strategy obviously only scales to s3 sizes where a download of all version IDs is doable in a reasonable amount of time. Maybe we should add a limit here for the maximum size we want to handle here. 50 thousand seems reasonable to me.

If the request returns 1k versions/deletions each time, then that would mean 50 requests.

I have done a local check, running `time aws s3api list-object-versions --bucket neon-github-ci-tests --max-items 100000 > test.json` and that already took 1 minute 32 seconds, and gave `100000` versions for various keys and `189811` delete markers, resulting in 200 MB of json.

With that speed, obtaining 10 million version markers would take 2 and a half hours and take up 20 GB.

So yeah, in the long run we'll probably need to create our own tracking of delete markers with the appropriate indices that allow us fast queries in them.

Reply:
Yeah, having a limit sounds reasonable to me, although it doesn't have to go in this pr imo.

Reply:
will add it to a future PR. made a more comprehensive comment in https://github.com/neondatabase/cloud/issues/8233#issuecomment-1915021625

Reply:
PR filed: #6515